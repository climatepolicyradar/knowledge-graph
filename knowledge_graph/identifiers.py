import hashlib
import re
from enum import Enum
from functools import total_ordering
from typing import Any, Callable

from pydantic import BaseModel, Field
from pydantic_core import CoreSchema, core_schema
from typing_extensions import Self


@total_ordering
class WikibaseID(str):
    """A Wikibase ID, which is a string that starts with a 'Q' followed by a number."""

    regex = r"^Q[1-9][0-9]*$"

    def __new__(cls, value):
        """Validate the Wikibase ID string and create a new instance."""
        cls._validate(value)
        return str.__new__(cls, value)

    @property
    def numeric(self) -> int:
        """The numeric value of the Wikibase ID"""
        return int(self[1:])

    def __lt__(self, other) -> bool:
        """Compare two Wikibase IDs numerically"""
        if isinstance(other, str):
            other = WikibaseID(other)
        if not isinstance(other, WikibaseID):
            return NotImplemented
        return self.numeric < other.numeric

    def __eq__(self, other) -> bool:
        """Check if two Wikibase IDs are equal"""
        if isinstance(other, str):
            other = WikibaseID(other)
        if not isinstance(other, WikibaseID):
            return NotImplemented
        return self.numeric == other.numeric

    def __hash__(self) -> int:
        """Hash a Wikibase ID consistently with string representation"""
        return hash(str(self))

    @classmethod
    def _validate(
        cls, __input_value: Any, _info: core_schema.ValidationInfo = None
    ) -> str:
        """Validate that the Wikibase ID is in the correct format."""
        if not isinstance(__input_value, str):
            raise ValueError(f"Wikibase ID must be a string, got {type(__input_value)}")
        if not re.match(cls.regex, __input_value):
            raise ValueError(f"'{__input_value}' is not a valid Wikibase ID")
        return __input_value

    @classmethod
    def __get_pydantic_core_schema__(
        cls, source_type: Any, handler: Callable[[Any], CoreSchema]
    ) -> CoreSchema:
        """Returns a pydantic_core.CoreSchema object for Pydantic V2 compatibility."""
        return core_schema.json_or_python_schema(
            json_schema=core_schema.str_schema(),
            python_schema=core_schema.with_info_plain_validator_function(cls._validate),
        )


class VespaSchema(Enum):
    """Schema definitions' names for Vespa"""

    FamilyDocument = "family_document"
    DocumentPassage = "document_passage"


class VespaID(BaseModel):
    """Base class for a Vespa schema ID"""

    prefix: str = Field(description="TODO", frozen=True, default="id:doc_search")
    kind: VespaSchema = Field(description="TODO")
    id: str = Field(description="TODO")

    def __str__(self) -> str:
        """String representation of a Vespa ID, ready to be used with Vespa queries"""
        return f"{self.prefix}:{self.kind.value}::{self.id}"


class FamilyDocumentID(VespaID):
    """An ID for a family document in Vespa"""

    kind: VespaSchema = Field(default=VespaSchema.FamilyDocument, exclude=True)


class DocumentPassageID(VespaID):
    """An ID for a document passage in Vespa"""

    kind: VespaSchema = Field(default=VespaSchema.DocumentPassage, exclude=True)


class Identifier(str):
    """
    A unique ID for a resource, comprised of 8 unambiguous lowercase and numeric characters.

    IDs are generated deterministically from input data, and look something like:
    ["2sgknw32", "gg7h2j2s", ...]

    With a set of 31 possible characters and 8 spaces in the ID, there's a total of
    31^8 = 852,891,037,441 available values in the space. This should be more than
    enough for most use cases!

    Usage:
      To generate an ID: `my_id = Identifier.generate("some", "data")`
      To cast/validate a string: `my_id = Identifier("abcdef12")` (raises ValueError if invalid)
    """

    # the following list of characters excludes "i", "l", "1", "o", "0" to minimise
    # ambiguity when people read the identifiers at a glance
    valid_characters = "abcdefghjkmnpqrstuvwxyz23456789"

    # Pattern needs to be defined using the literal string for valid_characters
    # as class attributes are resolved at class creation time.
    pattern = re.compile(rf"^[{valid_characters}]{{8}}$")

    def __new__(cls, value):
        """Validate the Identifier string and create a new instance."""
        cls._validate(value)
        return str.__new__(cls, value)

    @classmethod
    def generate(cls, *args) -> "Self":
        """Generates a new Identifier from the supplied data."""
        if not args:
            raise TypeError(
                f"{cls.__name__}.generate() requires at least one argument."
            )
        stringified_args = ""
        for arg in args:
            if isinstance(arg, BaseModel):
                stringified_args += arg.model_dump_json()
            else:
                stringified_args += str(arg)
        hashed_data = hashlib.sha256(stringified_args.encode()).digest()
        identifier = "".join(
            cls.valid_characters[b % len(cls.valid_characters)]
            for b in hashed_data[:8]  # Use first 8 bytes of hash
        )
        return cls(identifier)

    @classmethod
    def _validate(cls, value: str, _info: core_schema.ValidationInfo = None) -> str:
        """Validate that the Identifier string is in the correct format"""
        if not isinstance(value, str):
            raise TypeError(
                f"{cls.__name__} value must be a string, received {type(value).__name__}"
            )
        if not cls.pattern.match(value):
            raise ValueError(
                f"'{value}' is not a valid {cls.__name__}. Must be 8 characters from "
                f"the set '{cls.valid_characters}' (pattern: r'{cls.pattern.pattern}')."
            )
        return value

    @classmethod
    def __get_pydantic_core_schema__(
        cls, source_type: Any, handler: Callable[[Any], CoreSchema]
    ) -> CoreSchema:
        """Returns a pydantic_core.CoreSchema object for Pydantic V2 compatibility."""
        return core_schema.json_or_python_schema(
            json_schema=core_schema.str_schema(),
            python_schema=core_schema.with_info_plain_validator_function(cls._validate),
        )


class ClassifierID(Identifier):
    """
    A unique identifier specifically for classifiers.

    This is intended for typing clarity rather than extending the Identifier class.
    """

    pass
