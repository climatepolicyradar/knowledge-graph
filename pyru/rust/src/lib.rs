use pyo3::prelude::*;
use regex::Regex;
use serde_json::Value;
use std::collections::HashSet;

fn position_overlaps(matched_positions: &HashSet<usize>, start: usize, end: usize) -> bool {
    (start..end).any(|pos| matched_positions.contains(&pos))
}

fn predict_rust(
    text: &str,
    case_sensitive_pattern: &Regex,
    case_insensitive_pattern: &Regex,
) -> Vec<(usize, usize, String)> {
    let mut spans = Vec::new();
    let mut matched_positions = HashSet::new();

    // Case-sensitive matching
    for mat in case_sensitive_pattern.find_iter(text) {
        let (start, end) = (mat.start(), mat.end());
        if start != end && !position_overlaps(&matched_positions, start, end) {
            spans.push((start, end, text[start..end].to_string()));
            matched_positions.extend(start..end);
        }
    }

    // Case-insensitive matching
    for mat in case_insensitive_pattern.find_iter(text) {
        let (start, end) = (mat.start(), mat.end());
        if start != end && !position_overlaps(&matched_positions, start, end) {
            spans.push((start, end, text[start..end].to_string()));
            matched_positions.extend(start..end);
        }
    }

    spans
}

#[pyfunction]
fn batch_regex_predict(benchmark_data_json: &str) -> PyResult<(f64, usize, usize, usize)> {
    let start_time = std::time::Instant::now();
    
    // Parse the benchmark JSON data
    let benchmark_data: Value = serde_json::from_str(benchmark_data_json)
        .map_err(|e| pyo3::exceptions::PyValueError::new_err(format!("JSON parse error: {}", e)))?;
    let passages = benchmark_data["passages"].as_array().unwrap();
    let classifier_labels = &benchmark_data["classifier_labels"];
    
    // Extract texts
    let texts: Vec<&str> = passages
        .iter()
        .map(|p| p["text"].as_str().unwrap())
        .collect();
    
    // Build regex patterns from labels (outside timing)
    let case_sensitive_labels: Vec<&str> = classifier_labels["case_sensitive"]
        .as_array()
        .unwrap()
        .iter()
        .map(|v| v.as_str().unwrap())
        .collect();
    
    let case_insensitive_labels: Vec<&str> = classifier_labels["case_insensitive"]
        .as_array()
        .unwrap()
        .iter()
        .map(|v| v.as_str().unwrap())
        .collect();
    
    let cs_pattern_str = format!(r"\b(?:{})\b", case_sensitive_labels.join("|"));
    let ci_pattern_str = format!(r"(?i)\b(?:{})\b", case_insensitive_labels.join("|"));
    
    let case_sensitive_pattern = Regex::new(&cs_pattern_str).map_err(|e| {
        pyo3::exceptions::PyValueError::new_err(format!("Regex error: {}", e))
    })?;
    let case_insensitive_pattern = Regex::new(&ci_pattern_str).map_err(|e| {
        pyo3::exceptions::PyValueError::new_err(format!("Regex error: {}", e))
    })?;
    
    // Start timing after regex compilation
    let processing_start = std::time::Instant::now();
    
    let mut passages_with_matches = 0;
    let mut total_spans = 0;
    
    for text in &texts {
        let spans = predict_rust(text, &case_sensitive_pattern, &case_insensitive_pattern);
        if !spans.is_empty() {
            passages_with_matches += 1;
        }
        total_spans += spans.len();
    }
    
    let processing_time = processing_start.elapsed().as_secs_f64();
    let total_time = start_time.elapsed().as_secs_f64();
    
    println!("Rust regex compilation time: {:.4}s", total_time - processing_time);
    println!("Rust processing time: {:.4}s", processing_time);
    
    Ok((processing_time, passages_with_matches, total_spans, texts.len()))
}

#[pymodule]
fn pyru_rust(m: &Bound<'_, PyModule>) -> PyResult<()> {
    m.add_function(wrap_pyfunction!(batch_regex_predict, m)?)?;
    Ok(())
}