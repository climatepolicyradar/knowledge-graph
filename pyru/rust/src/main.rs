use regex::Regex;
use serde_json::Value;
use std::collections::HashSet;
use std::fs;
use std::time::Instant;

#[derive(Debug)]
struct Span {
    text: String,
    start_index: usize,
    end_index: usize,
}

#[derive(Debug)]
struct Passage {
    spans: Vec<Span>,
    has_matches: bool,
}

fn position_overlaps(matched_positions: &HashSet<usize>, start: usize, end: usize) -> bool {
    (start..end).any(|pos| matched_positions.contains(&pos))
}

fn predict(
    text: &str,
    case_sensitive_pattern: &Regex,
    case_insensitive_pattern: &Regex,
) -> Vec<Span> {
    let mut spans = Vec::new();
    let mut matched_positions = HashSet::new();

    // Case-sensitive matching
    for mat in case_sensitive_pattern.find_iter(text) {
        let (start, end) = (mat.start(), mat.end());
        if start != end && !position_overlaps(&matched_positions, start, end) {
            spans.push(Span {
                text: text[start..end].to_string(),
                start_index: start,
                end_index: end,
            });
            matched_positions.extend(start..end);
        }
    }

    // Case-insensitive matching
    for mat in case_insensitive_pattern.find_iter(text) {
        let (start, end) = (mat.start(), mat.end());
        if start != end && !position_overlaps(&matched_positions, start, end) {
            spans.push(Span {
                text: text[start..end].to_string(),
                start_index: start,
                end_index: end,
            });
            matched_positions.extend(start..end);
        }
    }

    spans
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("Loading benchmark data for Rust test...");

    // Load the benchmark data
    let benchmark_file = "../../benchmark_data_CCLW.executive.10014.4470.json";
    let benchmark_data_str = fs::read_to_string(benchmark_file)?;
    let benchmark_data: Value = serde_json::from_str(&benchmark_data_str)?;

    let passages = benchmark_data["passages"].as_array().unwrap();
    let classifier_labels = &benchmark_data["classifier_labels"];

    println!(
        "Loaded {} passages from {}",
        benchmark_data["total_blocks"],
        benchmark_data["document_id"].as_str().unwrap()
    );

    // Extract texts
    let texts: Vec<&str> = passages
        .iter()
        .map(|p| p["text"].as_str().unwrap())
        .collect();

    // Build regex patterns from labels
    let case_sensitive_labels: Vec<&str> = classifier_labels["case_sensitive"]
        .as_array()
        .unwrap()
        .iter()
        .map(|v| v.as_str().unwrap())
        .collect();

    let case_insensitive_labels: Vec<&str> = classifier_labels["case_insensitive"]
        .as_array()
        .unwrap()
        .iter()
        .map(|v| v.as_str().unwrap())
        .collect();

    // Create patterns
    let cs_pattern_str = format!(r"\b(?:{})\b", case_sensitive_labels.join("|"));
    let ci_pattern_str = format!(r"(?i)\b(?:{})\b", case_insensitive_labels.join("|"));

    let case_sensitive_pattern = Regex::new(&cs_pattern_str)?;
    let case_insensitive_pattern = Regex::new(&ci_pattern_str)?;

    println!(
        "Case sensitive labels: {} labels",
        case_sensitive_labels.len()
    );
    println!(
        "Case insensitive labels: {} labels",
        case_insensitive_labels.len()
    );

    // Single-threaded benchmark
    println!("\nRunning Rust single-threaded benchmark...");
    let start_time = Instant::now();

    let mut doc_labels = Vec::new();
    for text in &texts {
        let spans = predict(text, &case_sensitive_pattern, &case_insensitive_pattern);
        doc_labels.push(Passage {
            has_matches: !spans.is_empty(),
            spans,
        });
    }

    let total_time = start_time.elapsed().as_secs_f64();

    // Calculate metrics
    let passages_with_matches = doc_labels.iter().filter(|p| p.has_matches).count();
    let total_spans: usize = doc_labels.iter().map(|p| p.spans.len()).sum();

    println!("Rust single-threaded time: {:.4}s", total_time);
    println!("Passages with matches: {}", passages_with_matches);
    println!("Total spans found: {}", total_spans);
    println!("Total passages processed: {}", doc_labels.len());

    Ok(())
}
