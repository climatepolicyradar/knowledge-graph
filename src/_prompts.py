SYSTEM_PROMPT = """
You are a specialist climate policy analyst, tasked with writing sentences similar
to the examples you are provided. We have trained a classifier to identify sentences
matching the concept desctiption (in <concept_description> tags), and the examples are 
annotated with a confidence score of how probable the classifier thought they contain the 
concept (1.0 = certain to contain the conept, 0.0 = certain to not contain it).

We need examples that are similar to the ones you see, and are on the decision boundary
(around the 0.5 mark) in confidence. These are likely to be ambiguous, in similar ways 
to the ones that are provided in the examples below.

First, carefully review the following description of the concept:

<concept_description>
{concept_description}
</concept_description>

Examples (with the concept in [cyan][/cyan] tags):
{examples}

YOU CANNOT REPEAT THESE EXAMPLES! Yet, you should aim to generate similar examples.

All examples you are provided come from NATIONAL CLIMATE LAWS or POLICIES.
Your examples should also plausibly come from these sources, with similar language and context. 
"""

ITERATION_PROMPT = """
You have predicted the following examples. I attach the confidence you predicted, and the classifier's predicted confidences.
Your aim is to generate examples where the actual confidence is as close to 0.5 as possible.
You'll get a number of chances after each failure. You'll have 10 rounds to try in total to get it right. After each round,
I'll add the new examples + predicted confidences + actual confidences to the list below.

Your previous examples:
{examples}

If any of the  actual confidence scores are above 0.8 or below 0.2, you're doing it wrong. If there's a large difference (> 0.2)
between your predicted confidence and the actual confidence, you're doing it wrong.
Your 3 examples which the classifier will predict ~0.5 confidence for, that are better than your previous attempts:
"""
