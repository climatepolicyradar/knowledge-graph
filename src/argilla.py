from itertools import cycle
from typing import Any, Dict, Generator, Tuple

from argilla import FeedbackDataset, FeedbackRecord


def distribute_labelling_projects(
    datasets: list, labellers: list[str], min_labellers: int = 2
) -> Generator[Tuple[Any, str], None, None]:
    """
    Distribute labelling projects to labellers.

    For efficient labelling, tasks should be distributed such that each dataset is
    labelled by at least `min_labellers` labellers, and each labeller is assigned to a
    minimal number of datasets.

    :param list[] datasets: datasets to distribute among labellers
    :param list[str] labellers: list of labellers
    :param int min_labellers: minimum number of labellers per dataset, defaults to 2
    :return Generator[Tuple[Any, str], None, None]: a generator of tuples containing
        the dataset and the labeller assigned to it
    """
    if len(labellers) < min_labellers:
        raise ValueError(
            "number of items in labellers must be greater than or equal to min_labellers"
        )

    labeller_cycle = cycle(labellers)
    for dataset in datasets:
        for _ in range(min_labellers):
            yield dataset, next(labeller_cycle)


def combine_datasets(*datasets: FeedbackDataset) -> FeedbackDataset:  # type: ignore
    """
    Combine an arbitrary number of argilla datasets into one.

    :param FeedbackDataset *datasets: Unspecified number of datasets to combine, at
    least one.
    :return FeedbackDataset: The combined dataset
    """
    if not datasets:
        raise ValueError("At least one dataset must be provided")

    combined_dataset = FeedbackDataset(  # type: ignore
        fields=datasets[0].fields,  # type: ignore
        questions=datasets[0].questions,  # type: ignore
        metadata_properties=datasets[0].metadata_properties,  # type: ignore
        vectors_settings=datasets[0].vectors_settings,  # type: ignore
        guidelines=datasets[0].guidelines,  # type: ignore
        allow_extra_metadata=datasets[0].allow_extra_metadata,  # type: ignore
    )

    records_dict: Dict[str, FeedbackRecord] = {}  # type: ignore
    for dataset in datasets:
        for record in dataset.records:  # type: ignore
            # Use the 'text' field as the key (assuming it's unique)
            key = record.fields.get("text", "")

            if key in records_dict:
                # If the record already exists, merge the responses
                existing_record = records_dict[key]
                existing_record.responses.extend(record.responses)  # type: ignore
            else:
                # If it's a new record, add it to the dictionary
                records_dict[key] = record  # type: ignore

    # Convert the dictionary values back to a list of records
    combined_records = list(records_dict.values())

    # Add the combined records to the new dataset
    combined_dataset.add_records(combined_records)  # type: ignore

    return combined_dataset
