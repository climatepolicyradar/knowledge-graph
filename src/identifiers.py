import hashlib
import re


class WikibaseID(str):
    """A Wikibase ID, which is a string that starts with a 'Q' followed by a number."""

    regex = r"^Q[1-9]\d*$"

    @classmethod
    def _validate(cls, value: str, field=None) -> str:
        """Validate that the Wikibase ID is in the correct format"""
        if not re.match(cls.regex, value):
            raise ValueError(f"{value} is not a valid Wikibase ID")
        return value

    @classmethod
    def __get_validators__(cls):
        """Return a generator of validators for the WikibaseID class"""
        yield cls._validate

    def __new__(cls, value: str) -> "WikibaseID":
        """Create a new instance of WikibaseID after validation"""
        validated_value = cls._validate(value)
        return str.__new__(cls, validated_value)


def deterministic_hash(*args) -> int:
    """
    Generate a deterministic hash of the input data using SHA-256.

    Hashes should be consistent across different runs of a program.
    """
    input_string = "".join([str(arg) for arg in args])
    return int.from_bytes(
        hashlib.sha256(input_string.encode()).digest()[:8], byteorder="big"
    )


def generate_identifier(*args) -> str:
    """
    Generates a neat identifier using eight unambiguous lowercase and numeric characters

    The resulting identifiers look something like this: ["2sgknw32", "gg7h2j2s", ...]

    With a set of 31 possible characters and 8 positions, this function is able to
    generate 31^8 = 852,891,037,441 unique identifiers. This should be more than enough
    for most use cases!

    :param args: an arbitrary set of data to be hashed
    :return str: a unique identifier based on the input string
    """
    # the following list of characters excludes "i", "l", "1", "o", "0" to minimise
    # ambiguity when reading the identifiers
    characters = "abcdefghjkmnpqrstuvwxyz23456789"
    hashed_data = hashlib.sha256("".join([str(arg) for arg in args]).encode()).digest()
    return "".join(characters[b % len(characters)] for b in hashed_data[:8])
