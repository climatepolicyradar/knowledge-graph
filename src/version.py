import re


class Version:
    """A version as mandated by W&B."""

    regex = r"^v\d+$"

    value: int

    def __init__(self, value: str):
        """Create a new instance of Version after validation."""
        self.value = self._validate(value)

    @classmethod
    def _validate(cls, value: str) -> int:
        if value == "latest":
            raise ValueError("`latest` isn't yet supported")

        if not re.match(cls.regex, value):
            raise ValueError(
                'version must be in the format "v" followed by a number (e.g., v3)'
            )

        version_number = int(value[1:])

        min_version = 0

        if version_number < min_version:
            raise ValueError(
                f"version number must be greater than or equal to {min_version}"
            )

        return version_number

    def __str__(self):
        """Return a string representation of the Version."""
        return f"v{self.value}"

    def __repr__(self):
        """Return a string representation of the Version object."""
        return f"Version('v{self.value}')"

    def __eq__(self, other):
        """Check if this Version is equal to another Version or string."""
        if isinstance(other, Version):
            return self.value == other.value
        return str(self) == other

    def __lt__(self, other):
        """Check if this Version is less than another Version or string."""
        if isinstance(other, Version):
            return self.value < other.value
        return str(self) < other

    def __hash__(self):
        """Return a hash value for the Version."""
        return hash(self.value)

    def increment(self) -> "Version":
        """Increment the version number by 1."""
        return Version(f"v{self.value + 1}")
