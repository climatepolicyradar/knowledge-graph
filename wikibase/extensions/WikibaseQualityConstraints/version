WikibaseQualityConstraints: REL1_41
2024-02-14T07:25:46

9f52599
