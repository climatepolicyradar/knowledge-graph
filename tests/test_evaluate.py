from pathlib import Path
from unittest.mock import Mock, patch

import pandas as pd
import pytest
import typer
from syrupy.assertion import SnapshotAssertion
from wandb.wandb_run import Run

from scripts.evaluate import (
    Source,
    build_metrics_path,
    calculate_performance_metrics,
    create_gold_standard_labelled_passages,
    load_classifier_local,
    load_classifier_remote,
    log_metrics,
    print_metrics,
    save_metrics,
    validate_args,
    validate_local_args,
    validate_remote_args,
)
from src.identifiers import WikibaseID
from src.version import Version


def test_print_metrics(capsys, metrics_df: pd.DataFrame):
    print_metrics(metrics_df)

    captured = capsys.readouterr()

    # Assert the output contains some expected formatted values
    assert "0.99" in captured.out  # F1
    assert "0.3" in captured.out  # Precision
    assert "215" in captured.out  # Support


@pytest.mark.parametrize(
    "classifier,version,expected_error",
    [
        (None, None, None),
        (
            "TestClassifier",
            None,
            "classifier and version should not be specified",
        ),
        (
            None,
            Version("v1"),
            "classifier and version should not be specified",
        ),
        (
            "TestClassifier",
            Version("v1"),
            "classifier and version should not be specified",
        ),
    ],
)
def test_validate_local_args(classifier, version, expected_error):
    if expected_error is None:
        validate_local_args(classifier, version)  # Should not raise
    else:
        with pytest.raises(typer.BadParameter) as excinfo:
            validate_local_args(classifier, version)
        assert expected_error in str(excinfo.value)


@pytest.mark.parametrize(
    "track,classifier,version,expected_error",
    [
        (True, "TestClassifier", Version("v1"), None),
        (False, None, None, None),
        (
            False,
            "TestClassifier",
            None,
            "script was told not to track",
        ),
        (
            False,
            None,
            Version("v1"),
            "script was told not to track",
        ),
        (
            True,
            None,
            Version("v1"),
            "without a classifier name",
        ),
        (
            True,
            "TestClassifier",
            None,
            "without a version",
        ),
    ],
)
def test_validate_remote_args(track, classifier, version, expected_error):
    if expected_error is None:
        validate_remote_args(track, classifier, version)  # Should not raise
    else:
        with pytest.raises(typer.BadParameter) as excinfo:
            validate_remote_args(track, classifier, version)
        assert expected_error in str(excinfo.value)


@pytest.mark.parametrize(
    "track,classifier,version,expected_exception",
    [
        (True, None, None, None),
        (True, None, Version("v1"), pytest.raises(typer.BadParameter)),
        (True, "TestClassifier", None, pytest.raises(typer.BadParameter)),
        (True, "TestClassifier", Version("v1"), None),
        (False, None, Version("v1"), pytest.raises(typer.BadParameter)),
        (False, "TestClassifier", None, pytest.raises(typer.BadParameter)),
        (False, None, None, None),
        (False, "TestClassifier", Version("v1"), pytest.raises(typer.BadParameter)),
    ],
)
def test_validate_args(
    track: bool,
    classifier: str,
    version: Version,
    expected_exception: Exception,
    snapshot: SnapshotAssertion,
):
    source = Source.REMOTE if classifier and version else Source.LOCAL
    if expected_exception:
        # Instead of using the `match` arg on `pytest.raises`, this gives
        # a roundabout way for the snapshot to be used.
        with expected_exception as excinfo:
            validate_args(track, classifier, version, source)
        assert str(excinfo.value) == snapshot
    else:
        result = validate_args(track, classifier, version, source)
        assert result == snapshot


@pytest.mark.parametrize(
    "wikibase_id",
    [
        (WikibaseID("Q123")),
        (WikibaseID("Q330")),
    ],
)
def test_build_metrics_path(wikibase_id: str, snapshot: SnapshotAssertion):
    path = build_metrics_path(wikibase_id)
    # Convert to relative path for consistent snapshot testing across
    # environments
    relative_path = Path(
        *path.parts[-3:]
    )  # Take last 3 parts: classifiers_performance/Q123.json
    assert relative_path == snapshot


def test_save_metrics(tmp_path, metrics_df: pd.DataFrame):
    wikibase_id = WikibaseID("Q123")

    # Temporarily override metrics_dir
    import scripts.evaluate as evaluate

    original_metrics_dir = evaluate.metrics_dir
    evaluate.metrics_dir = tmp_path

    try:
        # Save metrics and verify
        metrics_path = save_metrics(metrics_df, wikibase_id)
        assert metrics_path.exists()
        assert metrics_path.suffix == ".json"

        # Load and verify contents
        loaded_df = pd.read_json(metrics_path, orient="records")
        pd.testing.assert_frame_equal(metrics_df, loaded_df)
    finally:
        # Restore original metrics_dir
        evaluate.metrics_dir = original_metrics_dir


@pytest.fixture
def mock_classifier():
    with patch("scripts.evaluate.Classifier") as mock:
        yield mock


def test_load_classifier_local(mock_classifier):
    with patch("scripts.evaluate.classifier_dir", Path("classifiers")):
        wikibase_id = WikibaseID("Q123")
        expected_path = Path("classifiers") / wikibase_id

        # Call function
        load_classifier_local(wikibase_id)

        # Verify Classifier.load was called with correct path
        mock_classifier.load.assert_called_once_with(expected_path)


def test_load_classifier_local_not_found(mock_classifier):
    with patch("scripts.evaluate.classifier_dir", Path("classifiers")):
        wikibase_id = WikibaseID("Q999")
        mock_classifier.load.side_effect = FileNotFoundError()

        with pytest.raises(typer.BadParameter) as exc_info:
            load_classifier_local(wikibase_id)

    assert "Classifier for Q999 not found" in str(exc_info.value)
    assert "just train Q999" in str(exc_info.value)


def test_load_classifier_remote(mock_classifier):
    # Setup
    run = Mock(spec=Run)
    run.config = {}  # Add dict for config
    classifier_name = "TestClassifier"
    version = Version("v1")
    wikibase_id = WikibaseID("Q123")

    # Mock the artifact
    mock_artifact = Mock()
    mock_artifact.metadata = {"aws_env": "test-env"}
    mock_artifact.download.return_value = "downloaded_path"
    run.use_artifact.return_value = mock_artifact

    # Call function
    result = load_classifier_remote(run, classifier_name, version, wikibase_id)

    # Verify artifact was retrieved correctly
    run.use_artifact.assert_called_once_with("Q123/TestClassifier:v1", type="model")

    # Verify classifier was loaded from downloaded path
    mock_classifier.load.assert_called_once_with(Path("downloaded_path/model.pickle"))

    # Verify AWS environment was set
    assert run.config["aws_env"] == "test-env"
    assert result == mock_classifier.load.return_value


def test_calculate_performance_metrics(concept, snapshot: SnapshotAssertion):
    # Create gold standard passages
    gold_standard = create_gold_standard_labelled_passages(concept)
    # Use the same passages as both gold and model predictions for testing
    result = calculate_performance_metrics(gold_standard, gold_standard)
    assert result == snapshot


def test_create_gold_standard_labelled_passages(concept, snapshot: SnapshotAssertion):
    result = create_gold_standard_labelled_passages(concept)
    assert result == snapshot


def test_log_metrics(metrics_df: pd.DataFrame):
    with patch("wandb.Table") as mock_wandb_table:
        mock_run = Mock()

        log_metrics(mock_run, metrics_df)

        mock_wandb_table.assert_called_once_with(
            data=metrics_df.values.tolist(),
            columns=metrics_df.columns.tolist(),
        )

        mock_run.log.assert_called_once()
        log_call = mock_run.log.call_args[0][0]
        assert "performance" in log_call
        assert log_call["performance"] == mock_wandb_table.return_value
