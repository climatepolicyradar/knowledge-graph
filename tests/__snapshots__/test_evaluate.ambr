# serializer version: 1
# name: test_validate_args[False-None-None-None]
  None
# ---
# name: test_validate_args[False-None-version4-expected_exception4]
  'A remote version (v1) was specified, but the script was told not to track. Tracking must be enabled to use a remote version.'
# ---
# name: test_validate_args[False-TestClassifier-None-expected_exception5]
  'A remote classifier (TestClassifier) was specified, but the script was told not to track. Tracking must be enabled to use a remote classifier.'
# ---
# name: test_validate_args[False-TestClassifier-version7-expected_exception7]
  'invalid evaluation arguments combination'
# ---
# name: test_validate_args[True-None-None-None]
  None
# ---
# name: test_validate_args[True-None-version1-expected_exception1]
  'cannot track a remote model artifact without a classifier name'
# ---
# name: test_validate_args[True-TestClassifier-None-expected_exception2]
  'cannot track a remote model artifact without a version'
# ---
# name: test_validate_args[True-TestClassifier-version3-None]
  None
# ---
